qiskit/algorithms/optimizers/suffix_averaging.py

```diff
@@ -0,0 +1,163 @@
+
+        self._n_params_suffix = n_params_suffix
+        self._optimizer = optimizer
+
+        self._circ_params: list[POINT] = []
+
+        original_callback = self._optimizer.callback
+        if isinstance(self._optimizer, (SPSA, QNSPSA)):
+
+            def callback(
+                nfev: int,
+                x_next: POINT,
+                fx_next: float,
+                stepsize: np.floating,
+                is_accepted: bool,
+            ):
+                self._circ_params.append(x_next)
+                if original_callback is not None:
+                    original_callback(nfev, x_next, fx_next, stepsize, is_accepted)
+
+        elif isinstance(self._optimizer, GradientDescent):
+
+            def callback(nfev: int, x: POINT, fun: float, stepsize: float):
+                self._circ_params.append(x)
+                if original_callback is not None:
+                    original_callback(nfev, x, fun, stepsize)
+
+        else:
+
+            def callback(x):
+                self._circ_params.append(x)
+                if original_callback is not None:
+                    return original_callback(x)
+                return None
+
+        self._optimizer.callback = callback
+        super().__init__()
+
+    def get_support_level(self):
+        """Return support level dictionary"""
+        return self._optimizer.get_support_level()
+
+    def _return_suffix_average(self) -> np.ndarray:
+
+        if isinstance(self._optimizer, NFT):
+            n_params = 1 if isinstance(self._circ_params[0], float) else len(self._circ_params[0])
+            n_iterates = int(len(self._circ_params))
+            n_repitition = int(len(self._circ_params) / n_params)
+            if n_repitition < self._n_params_suffix:
+                warnings.warn("The total number of iterations is less than n_params_suffix.")
+                averaged_param = np.zeros_like(self._circ_params[0])
+                for j in range(n_repitition):
+                    averaged_param += self._circ_params[n_iterates - n_params * j - 1]
+                averaged_param /= self._n_params_suffix
+                return averaged_param
+
+            averaged_param = np.zeros_like(self._circ_params[0])
+            for j in range(self._n_params_suffix):
+                averaged_param += self._circ_params[n_iterates - n_params * j - 1]
+            averaged_param /= self._n_params_suffix
+
+        else:
+            n_iterates = len(self._circ_params)
+
+            if n_iterates < self._n_params_suffix:
+                warnings.warn("The total number of iterations is less than n_params_suffix.")
+                return np.average(np.asarray(self._circ_params), axis=0)
+
+            averaged_param = np.zeros_like(self._circ_params[0])
+            for j in range(self._n_params_suffix):
+                averaged_param += self._circ_params[n_iterates - j - 1]
+            averaged_param /= self._n_params_suffix
+
+        return averaged_param
+
+    def minimize(
```

Comment:
We should make sure that this optimizer is reusable, so maybe `minimize` should reset `_circ_params`

Reply:
Thank you for the suggestion. I insert initialization of `_circ_params` at the beginning of `minimize`, [6a9e972](https://github.com/Qiskit/qiskit-terra/pull/8628/commits/6a9e972ef4ab4031b68b6ea3cb4028edc6f63f07).